### ISSUE TITLE
Uglify fails - Unknown provider: aProvider <- a

### ISSUE THREAD

**Issue by mastix** [angular-ui/bootstrap#2246]
Hi guys,

I ran into an issue yesterday when I've added the angular-bootstrap JS to my project (which has been set up using the Yeoman Angular Generator) and then it failed due to uglification. 

I've added the lib via bower (uncompressed!) and it got included into the vendor.js for concatination and minification.

When the file (_ui-bootstrap-tpls.js_; uncompressed) is concatenated and bundled in one big vendor.js, everything looks fine,  but as soon as the uglify task kicks in (with mangle: true = default), this fails (although the grunt ngmin task should rewrite the code to use Angular's inline annotation method):

The error message was pretty much this one:

``` javascript
Error: [$injector:unpr] Unknown provider: aProvider <- a
```

Pretty much like this bug: https://github.com/yeoman/generator-angular/issues/378

Is this a bug in the angular-bootstrap code and how it's implemented? 
Does anybody have an idea why this is happening and how it can be fixed (without turning mangle off)?

Just a hint would be cool! :+1: 

I've set _mangle_ to _false_ in the uglify.js code... and everything works fine, but I'd really like to turn it back to true again. :)

Regards,

mastix

**Comment by frhd** [comment]
@sethen Same here. What was the fix?

**Comment by dancancro** [comment]
I'm getting this too.  The app works in dev but not prod.  Unfortunately the proposed solutions aren't working for me.  

@PeterTrotter, I tried that but got the same results.
@DFOXpro, I have no uglify task in my Gruntfile whose mangle property I can set.  I think uglify is configured automatically by usemin.  I'm using the generator-angular-fullstack approach which does a lot of things I don't fully understand.

@knutesten, I'm confused how your idea solves the problem.  I want to use ngAnnotate to take care of minification-safe dependency injection, not do it myself.  That is, I don't want to have to write $scope and the other parameters twice everywhere as your solution implies unless I'm misunderstanding.  Is modalOptions something specific to your app?

The initial error isn't very useful.  So in dist/public/index.html, I replaced dist/app/d41d8cd9.vendor.js with the unminified-but-concatenated vendor.js and I changed the reference from angular.min.js to angular.js.

Having done that, I got the slightly more useful error message: Unknown provider: $urlMatcherFactory.  Any ideas why this one is a problem?  It's failing to find it in an array var called cache that does have these at the time of the exception being thrown
$$asyncCallbackProvider: Object
$$jqLiteProvider: Object
$$qProvider: Object
$$rAFProvider: Object
$$sanitizeUriProvider: Object
$$testabilityProvider: Object
$anchorScrollProvider: Object
$animateProvider: Object
$browserProvider: Object
$cacheFactoryProvider: Object
$compileProvider: Object
$controllerProvider: Object
$cookieStoreProvider: Object
$cookiesProvider: Object
$documentProvider: Object
$exceptionHandlerProvider: Object
$filterProvider: Object
$httpBackendProvider: Object
$httpProvider: Object
$injector: Object
$interpolateProvider: Object
$intervalProvider: Object
$localeProvider: Object
$locationProvider: Object
$logProvider: Object
$parseProvider: Object
$provide: Object
$qProvider: Object
$resolveProvider: Object
$resourceProvider: Object
$rootElementProvider: Object
$rootScopeProvider: Object
$sanitizeProvider: Object
$sceDelegateProvider: Object
$sceProvider: Object
$snifferProvider: Object
$templateCacheProvider: Object
$templateFactoryProvider: Object
$templateRequestProvider: Object
$timeoutProvider: Object
$urlMatcherFactory: Object
$urlMatcherFactoryProvider: Object
$windowProvider: Object
aDirectiveProvider: Object
currencyFilterProvider: Object
dateFilterProvider: Object
filterFilterProvider: Object
formDirectiveProvider: Object
inputDirectiveProvider: Object
jsonFilterProvider: Object
limitToFilterProvider: Object
linkyFilterProvider: Object
lowercaseFilterProvider: Object
maxlengthDirectiveProvider: Object
minlengthDirectiveProvider: Object
ngBindDirectiveProvider: Object
ngBindHtmlDirectiveProvider: Object
ngBindTemplateDirectiveProvider: Object
ngBlurDirectiveProvider: Object
ngChangeDirectiveProvider: Object
ngCheckedDirectiveProvider: Object
ngClassDirectiveProvider: Object
ngClassEvenDirectiveProvider: Object
ngClassOddDirectiveProvider: Object
ngClickDirectiveProvider: Object
ngCloakDirectiveProvider: Object
ngControllerDirectiveProvider: Object
ngCopyDirectiveProvider: Object
ngCutDirectiveProvider: Object
ngDblclickDirectiveProvider: Object
ngDisabledDirectiveProvider: Object
ngFocusDirectiveProvider: Object
ngFormDirectiveProvider: Object
ngHideDirectiveProvider: Object
ngHrefDirectiveProvider: Object
ngIfDirectiveProvider: Object
ngIncludeDirectiveProvider: Object
ngInitDirectiveProvider: Object
ngKeydownDirectiveProvider: Object
ngKeypressDirectiveProvider: Object
ngKeyupDirectiveProvider: Object
ngListDirectiveProvider: Object
ngMaxDirectiveProvider: Object
ngMaxlengthDirectiveProvider: Object
ngMinDirectiveProvider: Object
ngMinlengthDirectiveProvider: Object
ngModelDirectiveProvider: Object
ngModelOptionsDirectiveProvider: Object
ngMousedownDirectiveProvider: Object
ngMouseenterDirectiveProvider: Object
ngMouseleaveDirectiveProvider: Object
ngMousemoveDirectiveProvider: Object
ngMouseoutDirectiveProvider: Object
ngMouseoverDirectiveProvider: Object
ngMouseupDirectiveProvider: Object
ngNonBindableDirectiveProvider: Object
ngOpenDirectiveProvider: Object
ngOptionsDirectiveProvider: Object
ngPasteDirectiveProvider: Object
ngPatternDirectiveProvider: Object
ngPluralizeDirectiveProvider: Object
ngReadonlyDirectiveProvider: Object
ngRepeatDirectiveProvider: Object
ngRequiredDirectiveProvider: Object
ngSelectedDirectiveProvider: Object
ngShowDirectiveProvider: Object
ngSrcDirectiveProvider: Object
ngSrcsetDirectiveProvider: Object
ngStyleDirectiveProvider: Object
ngSubmitDirectiveProvider: Object
ngSwitchDefaultDirectiveProvider: Object
ngSwitchDirectiveProvider: Object
ngSwitchWhenDirectiveProvider: Object
ngTranscludeDirectiveProvider: Object
ngValueDirectiveProvider: Object
numberFilterProvider: Object
optionDirectiveProvider: Object
orderByFilterProvider: Object
patternDirectiveProvider: Object
requiredDirectiveProvider: Object
scriptDirectiveProvider: Object
selectDirectiveProvider: Object
socketFactoryProvider: Object
styleDirectiveProvider: Object
textareaDirectiveProvider: Object
uppercaseFilterProvider: Object

**Comment by yousfiSaad** [comment]
Thank you @knutesten, it solved my problem :)

**Comment by sethen** [comment]
@frhd Sorry, just seeing this now.

I can't remember exactly what the issue was anymore (I answered that almost a year ago)...  I remember it had something to do with the minification process in Yeoman somehow issuing a `grunt build:dist` I believe.  I think `ui-router` will throw this error when you're doing something like:

```javascript
.state('site.gateway.password-request', {
    url: '^/password/request',
    templateUrl: '/views/partials/templates/request.html',
    controller: function ($rootScope) {
        $rootScope.title = 'Password Request';
    }
})
```

This looks right, but you need to give the controller an array with your injected stuff first like:

```javascript
.state('site.gateway.password-request', {
    url: '^/password/request',
    templateUrl: '/views/partials/templates/request.html',
    controller: [$rootScope, function ($rootScope) {
        $rootScope.title = 'Password Request';
    }]
})
```

This may not help a lot, but that's all I can really remember.

**Comment by jonny2779** [comment]
For me it was the resolves in the states for ui router

```
$stateProvider.state('deals', {
      url: '/deals',
      templateUrl: '/contest/states/deals.template.html',
      controllerAs: 'DealsCtrl',
      resolve: {
        deals: ['$http', function ($http) {
          return $http.get('/api/deal/findactive')
            .then(function (results) {
              return results.data;
            });
        }],

```

**Comment by ryancarlson** [comment]
@mastix @fno @pkozlowski-opensource I just hit the same issue using ui-bootstrap with these tools.  The larger file size isn't an issue for me, so I'll work around it by not minifying that file.  Curious though.. did resolving this issue get set aside for one reason or another?

**Comment by aendrew** [comment]
I thought this issue was a thing of the past with ngAnnotate so was really confused when I ran into it again recently. Seems I removed the array syntax from a ui-router resolve — which ngAnnotate doesn't handle. From the Yeoman `generator-angular` Gruntfile:

> ngmin tries to make the code safe for minification automatically by using the Angular long form for dependency injection. It doesn't work on things like resolve or inject so those have to be done manually.

**Comment by Shaolin23** [comment]
@knutesten thanks - that was my oversight as well.

**Comment by glorat** [comment]
Amazing I found the solution to my problem in google leading here... @knutesten has my problem and solution (thanks!). minify is mucking up the of inline array controller syntax in the middle of code.

This won't happen if one declares the controller as a named controller rather than an inline one, which will rarely happen going forward since the documentation these days seems to promote that.

**Comment by ivomarsan** [comment]
Nice Idea @knutesten :smiley: but I solve my problem with 
`.pipe(uglify({
    mangle: false
}))`

__OBS:__ in gulp

**Comment by hc000** [comment]
@knutesten this solved my problem!

**Comment by icfantv** [comment]
Just a note for all of you here - you should ALWAYS turn on strictDi mode for your angular application.  You can do it either via the `ng-strict-di` attribute on your `ng-app` element or, if you're bootstrapping your application, you can follow the instructions [here](https://docs.angularjs.org/api/ng/function/angular.bootstrap).  This will protect you from these type of issues moving forward.

Unfortunately, not all angular examples are good about using the strict injection declaration format and so people learn incorrectly.

**Comment by hc000** [comment]
I just want to point out that for all other users even with using ng-strict-di, it did not resolve this problem, or lead to this specific case.

**Comment by derrickroccka** [comment]
I had a big problem with that too and I want to thank you all because a mix of answers was my good cocktail :D 

I had to uncomment uglify on the Gruntfile and add the option 'mangle' with the value 'false' as mentioned by @ivomarsan. 
I also made a separate build for "<script src="bower_components/angular-bootstrap/ui-bootstrap-tpls.js"></script>" and added the piece of code that @PeterTrotter said to the bower.json.

And that was the solution for me! 

Thank you so much!!

**Comment by mrdulin** [comment]
![image](https://cloud.githubusercontent.com/assets/17866683/14454239/0002ad1e-00cc-11e6-912d-f0fadd9d0107.png)

I found two problems,

The first is angular-bootstrap modal directive resolve problem, yeoman with grunt can not annotate the modal resolve dependency. so I write it manually.

But, The annotate problem is still there when I exec `grunt build`
![image](https://cloud.githubusercontent.com/assets/17866683/14454316/6c3de016-00cc-11e6-8360-e6e55214fa47.png)

I check the error stack in mangled code . I think the problem is angular-bootstrap modal, but i did't find the error code.

**Comment by joshribakoff** [comment]
This is a bug with ng-annotate, not ui-bootstrap. For me it started after upgrading ui.bootstrap though. If we look in the ng-annotate docs, we'll find:

> ng-annotate understands $uibModal.open (and $modal.open) (angular-ui/bootstrap).

There must be a bug with ng-annotate's handling of the newer ui.bootstrap, that uses the uib prefix. That is all I changed to replicate the bug was $modal.open to $uibModal.open, and that caused ng-annotate to stop implicitly matching.

If I tag the inline controller with @ngInject, it works:

```
var modalInstance = $uibModal.open({
                    /** @ngInject */
                    controller: function($scope, $uibModalInstance) {
```

In summary, this is a bug in ng-annotate, which is able to handle $modal.$open but apparently not $uibModal.open -- it is not in any way ui.bootstrap's problem.

**Comment by bennettatoms** [comment]
Hi, all. 
I was also having this issue on only one of several modal instances throughout my application, and after some debugging I realized that my problem was not using strict dependency injection in my resolve block, i.e.:

      var modalInstance = $uibModal.open({
        templateUrl: 'preferences.html',
        controller: 'preferencesCtrl as ctrl', // this external controller was using explicit function annotation...
        resolve: {
          parent: [function() { 
            return ctrl;
          }],
          sectorList: ['preferencesService', function(preferencesService) { // but this was not!
            return preferencesService.getSectors();
          }]
        }
      });

**Comment by northoutIshan** [comment]
Finally my issue is resolved. Please refer this url
https://stackoverflow.com/questions/17238759/angular-module-minification-bug

**Comment by shivamag91** [comment]
@knutesten thanks, your solution really helped a lot. I was stuck from a while n tried almost every possible  solution but not getting the reason behind. Now i get it . Thanks once again.

**Comment by jhonnyizidoro** [comment]
I'm using Laravel mix and fixed by inserting the config bellow
`mix.options({
	uglify: {
		uglifyOptions: {
			mangle: false,
		}
	}
})`

**Comment by jhonnyizidoro** [comment]
I'm using Laravel mix and fixed by inserting the config bellow

```javascript

mix.options({
  uglify: {
    uglifyOptions: {
      mangle: false,
    }
  }
})

```
